dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/expression/dataitem/ItemYearlyPeriodCount.java

```diff
@@ -0,0 +1,118 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.expression.dataitem;
+
+import static java.lang.String.valueOf;
+import static org.hisp.dhis.calendar.DateTimeUnit.fromJdkDate;
+import static org.hisp.dhis.parser.expression.ParserUtils.DOUBLE_VALUE_IF_NULL;
+import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.ExprContext;
+import static org.hisp.dhis.period.PeriodType.getPeriodFromIsoString;
+
+import java.util.List;
+
+import org.hisp.dhis.calendar.DateTimeUnit;
+import org.hisp.dhis.parser.expression.CommonExpressionVisitor;
+import org.hisp.dhis.parser.expression.ExpressionItem;
+import org.hisp.dhis.period.BiWeeklyAbstractPeriodType;
+import org.hisp.dhis.period.CalendarPeriodType;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.PeriodType;
+import org.hisp.dhis.period.WeeklyAbstractPeriodType;
+
+/**
+ * Expression item [yearlyPeriodCount]
+ *
+ * @author Jim Grace
+ */
+public class ItemYearlyPeriodCount
+    implements ExpressionItem
+{
+    @Override
+    public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+        return DOUBLE_VALUE_IF_NULL;
+    }
+
+    @Override
+    public Double evaluate( ExprContext ctx, CommonExpressionVisitor visitor )
+    {
+        List<Period> periods = visitor.getParams().getPeriods();
+
+        if ( periods.size() != 1 )
+        {
+            return 0.0; // Not applicable
+        }
+
+        Period period = periods.get( 0 );
+        PeriodType periodType = period.getPeriodType();
+
+        if ( periodType instanceof WeeklyAbstractPeriodType )
+        {
+            return weeksInYear( period );
+        }
+        else if ( periodType instanceof BiWeeklyAbstractPeriodType )
+        {
+            return biWeeksInYear( period );
+        }
+        else if ( periodType instanceof CalendarPeriodType && periodType.getFrequencyOrder() < 365 )
+        {
+            return Double.valueOf( ((CalendarPeriodType) periodType).generatePeriods( period ).size() );
+        }
+
+        return 1.0;
+    }
+
+    // -------------------------------------------------------------------------
+    // Supportive methods
+    // -------------------------------------------------------------------------
+
+    private double weeksInYear( Period period )
+    {
+        return testLastPeriodInYear( period, 52 );
```

Comment:
I think we could have some constants for some of the integers used in this class.
For someone reading the code, it's difficult to understand what `28` means. Why not `29` or `30`for example?
`12` could be assigned to a constant named `DECEMBER`, for example.

Reply:
Excellent comments, thanks! I've put all the magic numbers into constants and/or JavaDoc, and added to the JavaDoc so things will be easier for the next readers.

Also, as I was writing the JavaDoc I noticed that I'd not parsed `BiMonthly` periods correctly. I've fixed that now and added a test for it. I guess that's one of the benefits of writing JavaDoc!